File: 1304-longest-happy-string/1304-longest-happy-string.java
class Solution {
    public String longestDiverseString(int a, int b, int c) {
        HashMap<Character,Integer> frequency=new HashMap<>();
        if(a>0) frequency.put('a',a);
        if(b>0) frequency.put('b',b);
        if(c>0) frequency.put('c',c);
        System.out.println(frequency);
        PriorityQueue<Character> maxHeap=new PriorityQueue<>((p,q)->frequency.get(q)-frequency.get(p));
        maxHeap.addAll(frequency.keySet());
        System.out.println(maxHeap);
        String result="";
        while(!maxHeap.isEmpty()){
            char c1=maxHeap.poll();
            if(result.length()>=2 && result.charAt(result.length()-1)==c1 && result.charAt(result.length()-2)==c1){
                if(maxHeap.isEmpty())break;
                char c2=maxHeap.poll();
                result+=c2;
                frequency.put(c2,frequency.get(c2)-1);
                if(frequency.get(c2)>0) maxHeap.add(c2);
                maxHeap.add(c1);
            }
            else{
                result+=c1;
                frequency.put(c1,frequency.get(c1)-1);
                if(frequency.get(c1)>0) maxHeap.add(c1);
            }
        }
        return result;
    }
}

File: 1304-longest-happy-string/1304-longest-happy-string.py
class Solution:
    def longestDiverseString(self, a: int, b: int, c: int) -> str:
        freq=[]
        if a>0:
            freq.append(('a',a))
        if b>0:
            freq.append(('b',b))
        if c>0:
            freq.append(('c',c))
        max_heap=[(-val,key) for key,val in freq]
        result=""
        heapq.heapify(max_heap)
        print(max_heap)
        while max_heap:
            val1,char1=heapq.heappop(max_heap)
            if len(result)>=2 and result[-1]==char1 and result[-2]==char1:
                if not max_heap:
                    break
                val2,char2=heapq.heappop(max_heap)
                val2+=1
                result+=char2
                if val2<0:
                    heapq.heappush(max_heap,(val2,char2))
                heapq.heappush(max_heap,(val1,char1))
            else:
                val1+=1
                result+=char1
                if val1<0:
                    heapq.heappush(max_heap,(val1,char1))
        return result